db/log_reader.cc

```diff
@@ -511,10 +528,25 @@ unsigned int Reader::ReadPhysicalRecord(Slice* result, size_t* drop_size) {
           return kBadRecord;
         }
         if (uncompressed_size > 0) {
+          if (fragment_checksum != nullptr) {
+            XXH3_64bits_update(uncompress_hash_state_,
+                               uncompressed_buffer_.get(), uncompressed_size);
```

Comment:
What if we directly update `hash_state_` here instead of in `ReadRecord()`? We don't need to verify at the fragment level (the `return kBadRecord;`), do we?

Reply:
I think we have to do this for some subtle reason. I originally tried the approach of updating `hash_state_` in `ReadPhysicalRecord()`, but failed some test cases like [this one](https://github.com/facebook/rocksdb/blob/1e9bf25f61fa5b08fe2b3fe5c1fcf33e4c1c515c/db/log_test.cc#L504). If we have a sequence of two non-empty `kFirstType` records, then during `ReadPhysicalRecord()` of the second record, `hash_state_` will be tainted, and there is no way to recompute `hash_state_` in `ReadRecord()` based off the original buffer of uncompressed data.

Reply:
I see. `ReadPhysicalRecord()` could do `XXH3_64bits_reset()` for full types and first types. But it's probably too confusing; leaving it as it is sounds good to me.